Refactor clang_tidy.py

1ntEgr8 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61121
* **#61119**

Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared in the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

Differential Revision: [D29568582](https://our.internmc.facebook.com/intern/diff/D29568582)

1ntEgr8 committed:
Refactor clang_tidy.py

Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared into the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Refactor clang_tidy.py"


Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared into the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

[ghstack-poisoned]

samestep commented on mypy.ini:
```diff
@@ -38,7 +38,7 @@ files =
     test/test_type_info.py,
     test/test_utils.py,
     tools/linter/clang_format_utils.py,
-    tools/linter/clang_tidy.py,
+    tools/linter/clang_tidy,
     tools/generate_torch_version.py,
     tools/render_junit.py,
     tools/stats
```
we can probably just remove all of these lines since all of `tools/` is already included in `mypy-strict.ini`, right?

samestep commented on CONTRIBUTING.md:
````diff
@@ -1128,7 +1128,7 @@ have more checks than older versions. In our CI, we run clang-tidy-6.0.
    given revision:
   ```bash
-  python tools/linter/clang_tidy.py -d build -p torch/csrc --diff 'HEAD~1'
+  python tools/linter/clang_tidy -d build -p torch/csrc --diff 'HEAD~1'
````
when I run this, I get the following error:
```
clang_tidy: error: unrecognized arguments: -d
```

1ntEgr8 replied:
Good catch, the docs need to be updated. The flags have changed.

driazati replied:
Ah I forgot to update this in #60048, thanks for fixing it!

janeyx99 replied:
You probably already know this, but when I removed the -d flag, the next error was

`FileNotFoundError: [Errno 2] No such file or directory: 'HEAD~1'`

samestep commented on tools/git-pre-commit:
```diff
@@ -7,7 +7,7 @@ python tools/linter/flake8_hook.py
-  python tools/linter/clang_tidy.py \
+  python tools/linter/clang_tidy \
     --paths torch/csrc \
     --diff HEAD \
```
same issue here regarding "unrecognized arguments"

samestep reviewed (commented):
just to clarify: do any of the scripts use `requirements.txt`? or is it expected that the user will manually run `pip install -r tools/linter/clang_tidy/requirements.txt` first?

1ntEgr8 commented:
> just to clarify: do any of the scripts use `requirements.txt`? or is it expected that the user will manually run `pip install -r tools/linter/clang_tidy/requirements.txt` first?

It is expected that the user will run the pip install command themselves. I was debating if it was a good design choice to have a script that is supposed to execute a lint install the dependencies it needs. What do you think?

samestep commented:
> It is expected that the user will run the pip install command themselves. I was debating if it was a good design choice to have a script that is supposed to execute a lint install the dependencies it needs. What do you think?

Yeah I think it's fine if the lint script itself doesn't install the dependencies; you could add it to the `make setup_lint` target though, right?

driazati commented on tools/linter/clang_tidy/requirements.txt:
```diff
@@ -0,0 +1 @@
+unidiff==0.6.0
```
For a follow up: If we make the codebase clang-tidy clean and remove the line filter I don't think we'd need this anymore since it's pretty easy to manually parse out filenames from a diff file (as opposed to parsing names + hunk line ranges which isn't as straightforward)

driazati commented on tools/linter/clang_tidy/setup.sh:
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
+
+if [ ! -d build ]; then
```
While we're refactoring can we switch this to Python instead of mixing in Bash? Something like

```python
if not pathlib.Path("build").exists:
   subprocess.run([shutil.which("time"), sys.executable, ...])
   ...
```

1ntEgr8 replied:
Sure, we can do that

1ntEgr8 committed:
Update on "Refactor clang_tidy.py"


Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared in the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Refactor clang_tidy.py"


Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared in the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

[ghstack-poisoned]

driazati commented on Makefile:
```diff
@@ -70,6 +70,7 @@ setup_lint:
 		--job 'shellcheck' --step 'Install ShellCheck' --no-quiet; \
 	fi
 	pip install jinja2
+	pip install -r tools/linter/clang_tidy/requirements.txt
```
isn't this not necessary since we install it in the docker image?

https://github.com/pytorch/test-infra/blob/cb23e13ed41b4ae5cdf158e1fa9888bd0d8515f2/ci-docker-base/Dockerfile.cilint-clang-tidy#L35-L36

malfet replied:
But this is to run it on user system, isn't it?

driazati commented on Makefile:
```diff
@@ -116,4 +117,4 @@ toc:
 lint: flake8 mypy quick_checks cmakelint shellcheck
 
 quicklint: CHANGED_ONLY=--changed-only
-quicklint: mypy flake8 mypy quick_checks cmakelint shellcheck
```
nice catch

driazati commented on tools/linter/install/download_bin.py:
```diff
@@ -0,0 +1,164 @@
+
+    # This is the path to the file containing the reference hash.
+    hashpath = os.path.join(PYTORCH_ROOT, platform_to_hash[HOST_PLATFORM])
+    print(hashpath)
```
```suggestion
```

driazati reviewed (approved):
lgtm with some minor comments

1ntEgr8 commented on tools/linter/install/clang_tidy.py:
```diff
@@ -0,0 +1,18 @@
+}
+
+PLATFORM_TO_HASH = {
+    "Linux": os.path.join(HASH_PATH, "clang-tidy-linux64"),
```
I like the files approach because then its easy to write tools that automate updating the hash upon a new release (simply commit a modified version of the hash to the repo). If we inline it, it might make it a little more complex to do this (one approach would be to regex on the dict structure and update the hash, but then this makes our script coupled to the way we represent the OS to hash mapping)

1ntEgr8 committed:
Update on "Refactor clang_tidy.py"


Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared in the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

[ghstack-poisoned]

1ntEgr8 committed:
Update on "Refactor clang_tidy.py"


Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared in the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

[ghstack-poisoned]

1ntEgr8 commented:
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29568582).

1ntEgr8 committed:
Update on "Refactor clang_tidy.py"


Summary:

This change spilts the clang-tidy CI job into smaller steps and uses a
refactored version of the clang_tidy.py script.

The new folder structure is as follows:
```
tools/linter/clang_tidy
|_ __main__py
|_ requirements.txt
|_ run.py
|_ setup.sh
```

`__main__.py`

This script will run `tools/linter/clang_tidy/setup.sh` if a `build`
directory doesn't exist, mimicing what used to be done as a separate
step in the CI job.

After that, it will invoke `clang-tidy` with default arguments being
declared in the script itself (as opposed to declaring them in
lint.yml).

The reasoning behind this approach is two-fold:

- Make it easier to run `clang-tidy` locally using this script
- De-duplicate the option passing

`requirements.txt`

Contains a list of additional python dependencies needed by the
`clang-tidy` script.

`setup.sh`

If a build directory doesn't exist, this command will run the necessary
codegen and build commands for running `clang-tidy`

Example usage:
```
python3 tools/linter/clang_tidy --parallel
```
Notice that we don't have to put the `.py` at the end of `clang_tidy`.

Test Plan:

Run the following command:
```
python3 tools/linter/clang_tidy --paths torch/csrc/fx --parallel
```

Differential Revision: [D29568582](https://our.internmc.facebook.com/intern/diff/D29568582)

[ghstack-poisoned]

1ntEgr8 commented:
@1ntEgr8 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29568582).
